File: bench/dotnet/pingpong.cs
namespace Bench;
using System.Threading.Channels;
static class Pingpong
{
    // Two tasks bounce a token N round trips via channels. Checksum = N.
    public static async Task Run(int n)
    {
        var toW = Channel.CreateUnbounded<int>();
        var fromW = Channel.CreateUnbounded<int>();
        var worker = Task.Run(async () =>
        {
            while (true)
            {
                var m = await toW.Reader.ReadAsync();
                if (m < 0) break;
                await fromW.Writer.WriteAsync(m);
            }
        });
        int k = 0;
        while (k < n)
        {
            await toW.Writer.WriteAsync(k);
            await fromW.Reader.ReadAsync();
            k++;
        }
        await toW.Writer.WriteAsync(-1);
        await worker;
        Console.WriteLine(k);
    }
}

File: bench/clojure/pingpong.clj
;; Two threads bounce a token N round trips via blocking queues. Checksum = N.
(import '[java.util.concurrent LinkedBlockingQueue])
(let [n (long (Long/parseLong (or (System/getenv "BENCH_N") "100000")))
      q-to (LinkedBlockingQueue.)
      q-from (LinkedBlockingQueue.)
      worker (Thread. (fn [] (loop [] (let [m (.take q-to)] (when (>= (long m) 0) (.put q-from m) (recur))))))]
  (.start worker)
  (loop [k 0]
    (if (< k n)
      (do (.put q-to k) (.take q-from) (recur (inc k)))
      (do (.put q-to -1) (.join worker) (println k)))))
